<a href="https://colab.research.google.com/github/kbsmd-sportsmusicdata/Basketball-Analytics-Course/blob/main/pbpstats_gamelogs.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import requests

url = "https://api.pbpstats.com/get-game-logs/wnba"
params = {
    "Season": "2024", # Changed to a single season to troubleshoot 'Internal Server Error'
    "SeasonType": "Regular Season",
    "EntityId": "1611661319",
    "EntityType": "Team" # Use Opponent to get opponent stats
}
response = requests.get(url, params=params)

# Check if the request was successful before processing the JSON
if response.status_code == 200:
    response_json = response.json()
    totals = response_json['single_row_table_data']
    game_logs = response_json['multi_row_table_data']
    print("Data successfully retrieved.")
else:
    print(f"Error: API request failed with status code {response.status_code}")
    print(f"Response content: {response.text}")
    # Only try to parse as JSON if the content looks like JSON, or handle the JSONDecodeError
    try:
        error_json = response.json()
        print(f"Error JSON: {error_json}")
    except requests.exceptions.JSONDecodeError:
        print("Error response was not JSON formatted.")

Data successfully retrieved.


In [ ]:
display(game_logs[:5])

[{'Minutes': '40:00',
  'OffPoss': 84,
  'DefPoss': 83,
  'PenaltyOffPoss': 22,
  'PenaltyDefPoss': 13,
  'SecondChanceOffPoss': 10,
  'TotalPoss': 167,
  'AtRimFGM': 10,
  'AtRimFGA': 13,
  'SecondChanceAtRimFGM': 1,
  'SecondChanceAtRimFGA': 1,
  'PenaltyAtRimFGA': 1,
  'ShortMidRangeFGM': 8,
  'ShortMidRangeFGA': 24,
  'LongMidRangeFGM': 5,
  'LongMidRangeFGA': 5,
  'Corner3FGM': 1,
  'Corner3FGA': 8,
  'PenaltyCorner3FGA': 3,
  'Arc3FGM': 5,
  'Arc3FGA': 21,
  'SecondChanceArc3FGM': 1,
  'SecondChanceArc3FGA': 2,
  'PenaltyArc3FGA': 7,
  'FG2M': 23,
  'FG2A': 42,
  'FG3M': 6,
  'FG3A': 29,
  'FtPoints': 25,
  'Points': 89,
  'SecondChanceFG2M': 3,
  'SecondChanceFG2A': 5,
  'SecondChanceFG3M': 1,
  'SecondChanceFG3A': 2,
  'SecondChanceFtPoints': 4,
  'SecondChancePoints': 13,
  'PenaltyFG2M': 2,
  'PenaltyFG2A': 5,
  'PenaltyFG3A': 10,
  'PenaltyFtPoints': 16,
  'PenaltyPoints': 20,
  'PtsAssisted2s': 26,
  'PtsUnassisted2s': 20,
  'PtsAssisted3s': 12,
  'PtsUnassisted3s': 6,
  'P

In [ ]:
import requests

url = "https://api.pbpstats.com/get-game-logs/wnba"
params = {
    "Season": "2025", # Changed to a single season to troubleshoot 'Internal Server Error'
    "SeasonType": "Regular Season",
    "EntityId": "1611661319",
    "EntityType": "Team" # Use Opponent to get opponent stats
}
response = requests.get(url, params=params)

# Check if the request was successful before processing the JSON
if response.status_code == 200:
    response_json = response.json()
    totals = response_json['single_row_table_data']
    game_logs = response_json['multi_row_table_data']
    print("Data successfully retrieved.")
else:
    print(f"Error: API request failed with status code {response.status_code}")
    print(f"Response content: {response.text}")
    # Only try to parse as JSON if the content looks like JSON, or handle the JSONDecodeError
    try:
        error_json = response.json()
        print(f"Error JSON: {error_json}")
    except requests.exceptions.JSONDecodeError:
        print("Error response was not JSON formatted.")

Data successfully retrieved.


# Task
Fetch the 2024 WNBA regular season game logs for the team with EntityId "1611661319" and store them in a variable named `game_logs_2024`. Then, combine these 2024 logs with the previously fetched 2025 game logs (currently in the `game_logs` variable) into a single pandas DataFrame. Add a 'Season' column to differentiate between the two years, with '2024' for the 2024 data and '2025' for the 2025 data. Finally, save the combined DataFrame to a CSV file named `combined_game_logs.csv`.

## Fetch 2024 Game Logs

### Subtask:
Fetch the game logs for the 2024 season again and store them in a new variable named `game_logs_2024` to ensure both datasets are available for combination.


**Reasoning**:
To fetch the 2024 game logs as requested, I need to define the API endpoint, set the appropriate parameters for the 2024 season, make the API request, and then extract the relevant data into the `game_logs_2024` variable, including error handling.



In [ ]:
import requests

url = "https://api.pbpstats.com/get-game-logs/wnba"
params_2024 = {
    "Season": "2024",
    "SeasonType": "Regular Season",
    "EntityId": "1611661319",
    "EntityType": "Team"
}

response_2024 = requests.get(url, params=params_2024)

if response_2024.status_code == 200:
    response_json_2024 = response_2024.json()
    game_logs_2024 = response_json_2024['multi_row_table_data']
    print("2024 Game Logs successfully retrieved.")
else:
    print(f"Error: API request failed for 2024 season with status code {response_2024.status_code}")
    print(f"Response content: {response_2024.text}")
    try:
        error_json_2024 = response_2024.json()
        print(f"Error JSON: {error_json_2024}")
    except requests.exceptions.JSONDecodeError:
        print("Error response for 2024 season was not JSON formatted.")


2024 Game Logs successfully retrieved.


**Reasoning**:
Now that both 2024 and 2025 game logs are available as lists of dictionaries, I will convert them into pandas DataFrames, add a 'Season' column to each, concatenate them, and then save the combined DataFrame to a CSV file as specified in the task.



In [ ]:
import pandas as pd

# Convert 2024 game logs to DataFrame and add 'Season' column
df_2024 = pd.DataFrame(game_logs_2024)
df_2024['Season'] = '2024'

# Convert 2025 game logs to DataFrame and add 'Season' column
df_2025 = pd.DataFrame(game_logs)
df_2025['Season'] = '2025'

# Combine the two DataFrames
combined_game_logs = pd.concat([df_2024, df_2025], ignore_index=True)

# Save the combined DataFrame to a CSV file
combined_game_logs.to_csv('combined_game_logs.csv', index=False)

print("Combined game logs saved to 'combined_game_logs.csv'.")
display(combined_game_logs.head())

Combined game logs saved to 'combined_game_logs.csv'.


Minutes  OffPoss  DefPoss  PenaltyOffPoss  PenaltyDefPoss  \
0   40:00       84       83              22              13   
1   40:00       81       82              15               7   
2   40:00       82       82              15              14   
3   40:00       81       80              26              12   
4   40:00       79       78              10               6   

   SecondChanceOffPoss  TotalPoss  AtRimFGM  AtRimFGA  SecondChanceAtRimFGM  \
0                   10        167      10.0        13                   1.0   
1                   10        163       2.0         5                   NaN   
2                   10        164       1.0         3                   NaN   
3                    7        161       1.0         2                   1.0   
4                    9        157      13.0        27                   2.0   

   ...  3pt And 1 Free Throw Trips  Technical Free Throw Trips  \
0  ...                         NaN                         NaN   
1  ...                         NaN                         NaN   
2  ...                         NaN                         NaN   
3  ...                         NaN                         NaN   
4  ...                         NaN                         NaN   

   BlockedLongMidRange  Arc3PctBlocked  Transition Take Fouls Drawn  \
0                  NaN             NaN                          NaN   
1                  NaN             NaN                          NaN   
2                  NaN             NaN                          NaN   
3                  NaN             NaN                          NaN   
4                  NaN             NaN                          NaN   

   Charge Fouls  Transition Take Fouls  Charge Fouls Drawn  Season  \
0           NaN                    NaN                 NaN    2024   
1           NaN                    NaN                 NaN    2024   
2           NaN                    NaN                 NaN    2024   
3           NaN                    NaN                 NaN    2024   
4           NaN                    NaN                 NaN    2024   

   3SecondViolations  
0                NaN  
1                NaN  
2                NaN  
3                NaN  
4                NaN  

[5 rows x 227 columns]

## Final Task

### Subtask:
Confirm that the combined dataset has been successfully created and saved as a CSV file, and provide any further insights if requested.


## Summary:

### Q&A
Yes, the combined dataset has been successfully created and saved as `combined_game_logs.csv`.

### Data Analysis Key Findings
*   The 2024 WNBA regular season game logs for the team with `EntityId` "1611661319" were successfully fetched from the API with a 200 status code.
*   The 2024 game logs and the existing 2025 game logs were successfully converted into pandas DataFrames.
*   A new 'Season' column was added to each DataFrame, accurately labeling the data as '2024' or '2025'.
*   Both DataFrames were successfully concatenated into a single DataFrame named `combined_game_logs`.
*   The final `combined_game_logs` DataFrame was saved to a CSV file named `combined_game_logs.csv`.

### Insights or Next Steps
*   The combined dataset is now available for comparative analysis between the 2024 and 2025 WNBA regular seasons for the specified team.
*   Further analysis could involve exploring statistical differences or trends in team performance across the two seasons.


**bold text**

**get totals**: 2024 2025

In [ ]:
import requests

url = "https://api.pbpstats.com/get-totals/wnba"
params = {
    "Season": "2024", # Changed to a single season to troubleshoot 'Internal Server Error'
    "SeasonType": "Regular Season",
    "Type": "Team", # Corrected: 'Type' should be 'Team' or other permitted enumeration
    "TeamId": "1611661319" # Corrected: 'TeamId' should be the actual entity ID
}
response = requests.get(url, params=params)

# Check if the request was successful before processing the JSON
if response.status_code == 200:
    response_json = response.json()
    totals = response_json['single_row_table_data']
    # The get-totals endpoint typically returns only single_row_table_data for aggregated totals,
    # not multi_row_table_data (game logs). So, 'game_logs' variable might not be applicable here.
    # For this endpoint, we usually only expect 'single_row_table_data'.
    game_logs = response_json.get('multi_row_table_data', None) # Use .get() to avoid KeyError if not present
    print("Data successfully retrieved.")
else:
    print(f"Error: API request failed with status code {response.status_code}")
    print(f"Response content: {response.text}")
    # Only try to parse as JSON if the content looks like JSON, or handle the JSONDecodeError
    try:
        error_json = response.json()
        print(f"Error JSON: {error_json}")
    except requests.exceptions.JSONDecodeError:
        print("Error response was not JSON formatted.")

Error: API request failed with status code 500
Response content: {"detail":"Internal Server Error"}
Error JSON: {'detail': 'Internal Server Error'}


In [ ]:
import requests



url = "https://api.pbpstats.com/get-totals/wnba"
params = {
    "Season": "2024,2025",
    "SeasonType": "Regular Season",
    "Type": "Team" # Use Opponent for opponent stats
}
response = requests.get(url, params=params)
response_json = response.json()
league_stats = response_json["single_row_table_data"]
team_stats = response_json["multi_row_table_data"]url = "https://api.pbpstats.com/get-totals/nba"
params = {
    "Season": "2024-2025",
    "SeasonType": "Regular Season",
    "Type": "Team" # Use Opponent for opponent stats
}
response = requests.get(url, params=params)
response_json = response.json()
league_stats = response_json["single_row_table_data"]
team_stats = response_json["multi_row_table_data"]

In [ ]:
import requests

url = "https://api.pbpstats.com/get-totals/wnba"
params = {
    "Season": "2024",
    "SeasonType": "Regular Season",
    "Type": "Team",
    "TeamId": "1611661319" # Added TeamId for 'Type': 'Team'
}

response = requests.get(url, params=params)

# Check if the request was successful before processing the JSON
if response.status_code == 200:
    response_json = response.json()
    league_stats = response_json.get("single_row_table_data")
    team_stats = response_json.get("multi_row_table_data")
    print("Data successfully retrieved.")
    if league_stats:
        print("League Stats retrieved.")
    if team_stats:
        print("Team Stats retrieved.")
else:
    print(f"Error: API request failed with status code {response.status_code}")
    print(f"Response content: {response.text}")
    try:
        error_json = response.json()
        print(f"Error JSON: {error_json}")
    except requests.exceptions.JSONDecodeError:
        print("Error response was not JSON formatted.")

Data successfully retrieved.
League Stats retrieved.
Team Stats retrieved.


In [ ]:
import requests

url = "https://api.pbpstats.com/get-totals/wnba"
params = {
    "Season": "2025",
    "SeasonType": "Regular Season",
    "Type": "Team",
    "TeamId": "1611661319" # Added TeamId for 'Type': 'Team'
}

response = requests.get(url, params=params)

# Check if the request was successful before processing the JSON
if response.status_code == 200:
    response_json = response.json()
    league_stats = response_json.get("single_row_table_data")
    team_stats = response_json.get("multi_row_table_data")
    print("Data successfully retrieved.")
    if league_stats:
        print("League Stats retrieved.")
    if team_stats:
        print("Team Stats retrieved.")
else:
    print(f"Error: API request failed with status code {response.status_code}")
    print(f"Response content: {response.text}")
    try:
        error_json = response.json()
        print(f"Error JSON: {error_json}")
    except requests.exceptions.JSONDecodeError:
        print("Error response was not JSON formatted.")

Data successfully retrieved.
League Stats retrieved.
Team Stats retrieved.


# Task
Fetch the 2024 WNBA regular season team statistics using the `get-totals` API, combine them with the 2025 team statistics (already in the `team_stats` variable), add a 'Season' column to differentiate between the years, and save the combined data to a CSV file named `combined_team_stats.csv`.

## Fetch 2024 Team Stats

### Subtask:
Make an API request to `get-totals` for the 2024 regular season team statistics and store the results in a new variable, `team_stats_2024`.


**Reasoning**:
To fetch the 2024 regular season team statistics, I need to make an API request to the `get-totals` endpoint with the specified parameters, handle the response, and store the team statistics in `team_stats_2024`.



In [ ]:
import requests

url = "https://api.pbpstats.com/get-totals/wnba"
params = {
    "Season": "2024",
    "SeasonType": "Regular Season",
    "Type": "Team",
    "TeamId": "1611661319"
}

response = requests.get(url, params=params)

if response.status_code == 200:
    response_json = response.json()
    team_stats_2024 = response_json.get('multi_row_table_data', [])
    print("2024 Team Stats successfully retrieved.")
else:
    print(f"Error: API request failed for 2024 team stats with status code {response.status_code}")
    print(f"Response content: {response.text}")
    try:
        error_json = response.json()
        print(f"Error JSON: {error_json}")
    except requests.exceptions.JSONDecodeError:
        print("Error response for 2024 team stats was not JSON formatted.")

2024 Team Stats successfully retrieved.


**Reasoning**:
Now that both 2024 and 2025 team statistics are available as lists of dictionaries, I will convert them into pandas DataFrames, add a 'Season' column to each, concatenate them, and then save the combined DataFrame to a CSV file as specified in the task.



In [ ]:
import pandas as pd

# Convert 2024 team stats to DataFrame and add 'Season' column
df_team_stats_2024 = pd.DataFrame(team_stats_2024)
df_team_stats_2024['Season'] = '2024'

# Convert 2025 team stats to DataFrame and add 'Season' column
df_team_stats_2025 = pd.DataFrame(team_stats)
df_team_stats_2025['Season'] = '2025'

# Combine the two DataFrames
combined_team_stats = pd.concat([df_team_stats_2024, df_team_stats_2025], ignore_index=True)

# Save the combined DataFrame to a CSV file
combined_team_stats.to_csv('combined_team_stats.csv', index=False)

print("Combined team statistics saved to 'combined_team_stats.csv'.")
display(combined_team_stats.head())

Combined team statistics saved to 'combined_team_stats.csv'.


EntityId      TeamId Name ShortName       RowId TeamAbbreviation  \
0  1611661319  1611661319  LVA       LVA  1611661319              LVA   
1  1611661319  1611661319  LVA       LVA  1611661319              LVA   

   SecondsPlayed  GamesPlayed  PlusMinus  OffPoss  ...  ShootingFoulsDrawnPct  \
0        93900.0           39        215     3125  ...               0.108425   
1       105600.0           44        128     3436  ...               0.106613   

   TwoPtShootingFoulsDrawnPct  ThreePtShootingFoulsDrawnPct  \
0                    0.163079                      0.010597   
1                    0.156100                      0.017331   

   SecondChancePointsPct  PenaltyPointsPct  Avg2ptShotDistance  \
0               0.091587          0.193317            7.948342   
1               0.128261          0.244022            8.326892   

   Avg3ptShotDistance  PenaltyOffPossPct  Season  3SecondViolations  
0           24.856907           0.200320    2024                NaN  
1           24.877357           0.233993    2025                4.0  

[2 rows x 233 columns]

## Final Task

### Subtask:
Confirm the successful creation and saving of the `combined_team_stats.csv` file, and provide any further insights if requested.


## Summary:

### Q&A
Yes, the `combined_team_stats.csv` file was successfully created and saved.

### Data Analysis Key Findings
*   The 2024 WNBA regular season team statistics were successfully retrieved from the API with a status code of 200.
*   A 'Season' column was added to both the 2024 and 2025 team statistics DataFrames, allowing differentiation by year.
*   The 2024 and 2025 team statistics were successfully combined into a single DataFrame.
*   The combined data was saved to a CSV file named `combined_team_stats.csv`.

### Insights or Next Steps
*   The combined dataset is now available for direct year-over-year performance comparisons for the specified WNBA team.
*   Future analysis could involve fetching statistics for additional seasons or other team IDs to build a more comprehensive historical database.


In [ ]:
import requests
import pandas as pd
import json
from pathlib import Path
import time

# ============================================================================
# CONFIGURATION
# ============================================================================

# Team configuration
TEAM_ID_WNBA = "1611661319"  # Las Vegas Aces pbpstats team ID
TEAM_ABBREV = "LV"
SEASONS = ["2024"]  # WNBA uses single year format

# Data paths
DATA_DIR = Path("data")
RAW_DIR = DATA_DIR / "raw" / "pbpstats"
OUTPUT_DIR = DATA_DIR / "master"

RAW_DIR.mkdir(parents=True, exist_ok=True)
OUTPUT_DIR.mkdir(parents=True, exist_ok=True)

# API configuration
BASE_URL = "https://api.pbpstats.com"

print("="*70)
print("PBPSTATS SHOT DATA PULL (PYTHON)")
print("="*70)
print(f"Team: {TEAM_ABBREV}")
print(f"Seasons: {', '.join(SEASONS)}")
print(f"Data: Shot-by-shot locations and context")
print("="*70)
print()


# ============================================================================
# API FUNCTIONS
# ============================================================================

def make_api_request(endpoint, params):
    """Make GET request to pbpstats API."""
    url = f"{BASE_URL}{endpoint}"

    try:
        response = requests.get(url, params=params, timeout=30)
        response.raise_for_status()
        return response.json()
    except requests.exceptions.RequestException as e:
        print(f"  ✗ API Error: {e}")
        return None


def get_team_shots(season, team_id):
    """
    Get all shots for a team in a season.

    Returns shot-level data with:
    - x, y coordinates (court position)
    - shot_distance, shot_type (Arc3, Corner3, AtRim, etc.)
    - made (True/False)
    - shot_value (2 or 3 points)
    - player, lineup, opponent lineup
    - game context (quarter, time, score margin)
    - shot_quality (expected value 0-1)
    - blocked, assisted, putback indicators
    """
    print(f"\n[{season}] Fetching all team shots...")

    endpoint = "/get-shots/wnba"
    params = {
        "Season": season,
        "SeasonType": "Regular Season",
        "EntityType": "Team",
        "EntityId": "1611661319",
        # Optional filters:
        # "ShotType": "Arc3,Corner3"  # Filter to just 3s
        # "Outcome": "Made"  # Filter to makes only
        # "PlayerId": "204324"  # Filter to specific player
    }

    data = make_api_request(endpoint, params)

    if data is None or 'results' not in data or not data['results']:
        print(f"  ⚠ No shots found or invalid data structure")
        return pd.DataFrame()

    # Save raw response
    raw_file = RAW_DIR / f"shots_{season}_raw.json"
    with open(raw_file, 'w') as f:
        json.dump(data, f, indent=2)
    print(f"  → Saved raw: {raw_file}")

    # The response is a list of shot objects, nested under 'results' key
    # Convert to dataframe
    df = pd.DataFrame(data['results'])

    print(f"  ✓ Loaded {len(df)} shots")

    # Add season column
    df['season'] = season

    return df


# ============================================================================
# PROCESSING FUNCTIONS
# ============================================================================

def process_shot_data(df):
    """
    Clean and process shot data.
    Add shot zones, calculate shot quality tiers.
    """
    print(f"\n[PROCESSING] Cleaning shot data...")

    if len(df) == 0:
        return df

    # Ensure numeric columns
    numeric_cols = ['x', 'y', 'shot_distance', 'shot_value', 'score_margin',
                    'shot_quality', 'period', 'seconds_since_oreb',
                    'shot_time', 'start_time', 'end_time']

    for col in numeric_cols:
        if col in df.columns:
            df[col] = pd.to_numeric(df[col], errors='coerce')

    # Ensure boolean columns
    boolean_cols = ['made', 'blocked', 'assisted', 'putback']

    for col in boolean_cols:
        if col in df.columns:
            df[col] = df[col].astype(bool)

    # Create shot categories
    if 'shot_type' not in df.columns:
        if 'ShotType' in df.columns:
            df = df.rename(columns={'ShotType': 'shot_type'})
            print("  Renamed 'ShotType' to 'shot_type'.")
        else:
            print(f"  Error: Critical column 'shot_type' (or 'ShotType') not found. Cannot create shot categories. Available columns: {df.columns.tolist()}")
            # If 'shot_type' is essential for subsequent steps, you might raise an error or skip further processing.
            # For now, we'll return the dataframe without 'shot_category'.
            return df

    df['shot_category'] = df['shot_type'].map({
        'Arc3': 'Three',
        'Corner3': 'Three',
        'AtRim': 'Rim',
        'ShortMidRange': 'Short Mid',
        'LongMidRange': 'Long Mid'
    }).fillna('Other')

    # Shot quality tiers
    df['shot_quality_tier'] = pd.cut(
        df['shot_quality'],
        bins=[0, 0.35, 0.40, 0.50, 1.0],
        labels=['Poor', 'Average', 'Good', 'Excellent'],
        include_lowest=True
    )

    # Game situations
    df['clutch'] = (df['period'] >= 4) & (df['score_margin'].abs() <= 5)

    # Points scored
    df['points_scored'] = df.apply(
        lambda row: row['shot_value'] if row['made'] else 0,
        axis=1
    )

    # Shot zone (based on coordinates if needed)
    # x, y are in some coordinate system - might need court dimensions

    print(f"  ✓ Processed {len(df)} shots")
    print(f"  Shot types: {df['shot_category'].value_counts().to_dict()}")

    return df


def generate_shot_summary(df):
    """Generate comprehensive shot summary statistics."""
    print("\n" + "="*70)
    print("SHOT DATA SUMMARY")
    print("="*70)

    if len(df) == 0:
        print("No data to summarize")
        return

    print(f"\nTotal shots: {len(df)}")
    print(f"Made: {df['made'].sum()} ({df['made'].mean()*100:.1f}%)")

    # By shot type
    print("\n" + "-"*70)
    print("SHOT DISTRIBUTION")
    print("-"*70)

    # Ensure 'shot_category' exists before grouping
    if 'shot_category' in df.columns:
        shot_dist = df.groupby('shot_category').agg({
            'made': ['count', 'sum', 'mean']
        }).round(3)
        shot_dist.columns = ['Attempts', 'Makes', 'FG%']
        shot_dist['FG%'] = (shot_dist['FG%'] * 100).round(1)
        shot_dist = shot_dist.sort_values('Attempts', ascending=False)
        print(shot_dist)
    else:
        print("  Skipping shot distribution summary: 'shot_category' column not available.")


    # By shot quality
    if 'shot_quality_tier' in df.columns:
        print("\n" + "-"*70)
        print("SHOT QUALITY DISTRIBUTION")
        print("-"*70)

        quality_dist = df.dropna(subset=['shot_quality_tier']).groupby('shot_quality_tier').agg({
            'made': ['count', 'sum', 'mean'],
            'shot_quality': 'mean'
        }).round(3)
        quality_dist.columns = ['Attempts', 'Makes', 'FG%', 'Avg xFG']
        quality_dist['FG%'] = (quality_dist['FG%'] * 100).round(1)
        quality_dist['Avg xFG'] = (quality_dist['Avg xFG'] * 100).round(1)
        print(quality_dist)

    # Clutch shots
    if 'clutch' in df.columns:
        clutch_shots = df[df['clutch'] == True]
        if len(clutch_shots) > 0:
            print("\n" + "-"*70)
            print("CLUTCH SHOTS (Q4, close game)")
            print("-"*70)
            print(f"Attempts: {len(clutch_shots)}")
            print(f"Makes: {clutch_shots['made'].sum()}")
            print(f"FG%: {clutch_shots['made'].mean()*100:.1f}%")

    # Top shooters
    if 'player' in df.columns:
        print("\n" + "-"*70)
        print("TOP 5 SHOOTERS BY ATTEMPTS")
        print("-"*70)

        top_shooters = df.groupby('player').agg({
            'made': ['count', 'sum', 'mean']
        }).round(3)
        top_shooters.columns = ['Attempts', 'Makes', 'FG%']
        top_shooters['FG%'] = (top_shooters['FG%'] * 100).round(1)
        top_shooters = top_shooters.sort_values('Attempts', ascending=False).head(5)
        print(top_shooters)

    print("="*70)


# ============================================================================
# MAIN EXECUTION
# ============================================================================

def main():
    print(f"\nStarted: {pd.Timestamp.now()}\n")

    all_shots = []

    for season in SEASONS:
        print("="*70)
        print(f"PROCESSING SEASON {season}")
        print("="*70)

        # Fetch shots
        shots = get_team_shots(season, TEAM_ID_WNBA)

        if len(shots) > 0:
            # Process
            shots = process_shot_data(shots)

            # Append
            all_shots.append(shots)

    # Combine seasons
    if all_shots:
        print("\n" + "="*70)
        print("COMBINING SEASONS")
        print("="*70)

        shots_combined = pd.concat(all_shots, ignore_index=True)

        # Save to master
        parquet_file = OUTPUT_DIR / "aces_shots_pbpstats.parquet"
        shots_combined.to_parquet(parquet_file, index=False)
        print(f"\n✓ Saved: {parquet_file} ({len(shots_combined)} shots)")

        # Also save as CSV for easy viewing
        csv_file = OUTPUT_DIR / "aces_shots_pbpstats.csv"
        shots_combined.to_csv(csv_file, index=False)
        print(f"✓ Saved: {csv_file}")

        # Generate summary
        generate_shot_summary(shots_combined)

    else:
        print("\n⚠ No shot data loaded")

    print("\n" + "="*70)
    print("NEXT STEPS")
    print("="*70)
    print("""
1. Create shot charts in Tableau:
   - Load aces_shots_pbpstats.parquet or .csv
   - Plot x, y coordinates on court template
   - Color by made/missed, shot_type, or shot_quality

2. Analyze shot selection:
   - Compare shot distribution to league averages
   - Identify high-quality shot zones
   - Find underutilized efficient zones

3. Player shot profiles:
   - Filter by player
   - Show preferred shooting zones
   - Calculate effective FG% by zone

4. Lineup analysis:
   - Filter by lineup_id
   - Compare shot quality across lineups
   - Find lineups that generate best looks

AVAILABLE COLUMNS:
- Coordinates: x, y, shot_distance
- Type: shot_type, shot_category, shot_value
- Outcome: made, blocked, assisted, putback
- Context: player, lineup_id, opponent_lineup_id
- Game: gid (game_id), game_date, period, time
- Situation: score_margin, clutch, possession info
- Quality: shot_quality, shot_quality_tier
- Video: url (link to NBA video of shot)

SHOT TYPES:
- Arc3: Three-pointer above the break
- Corner3: Three-pointer from corners
- AtRim: Shots at the rim
- ShortMidRange: Mid-range closer to basket
- LongMidRange: Mid-range farther from basket

COORDINATE SYSTEM:
- x, y are in NBA court coordinates
- (0, 0) is center court
- Court dimensions: ~470 x ~250 units
- Positive y is towards one basket
    """)
    print("="*70)

    print(f"\nCompleted: {pd.Timestamp.now()}")


if __name__ == "__main__":
    main()

PBPSTATS SHOT DATA PULL (PYTHON)
Team: LV
Seasons: 2024
Data: Shot-by-shot locations and context


Started: 2026-01-07 06:11:41.437389

PROCESSING SEASON 2024

[2024] Fetching all team shots...
  → Saved raw: data/raw/pbpstats/shots_2024_raw.json
  ✓ Loaded 500 shots

[PROCESSING] Cleaning shot data...
  ✓ Processed 500 shots
  Shot types: {'Three': 192, 'Short Mid': 126, 'Rim': 115, 'Long Mid': 67}

COMBINING SEASONS

✓ Saved: data/master/aces_shots_pbpstats.parquet (500 shots)
✓ Saved: data/master/aces_shots_pbpstats.csv

SHOT DATA SUMMARY

Total shots: 500
Made: 222 (44.4%)

----------------------------------------------------------------------
SHOT DISTRIBUTION
----------------------------------------------------------------------
               Attempts  Makes   FG%
shot_category                       
Three               192     76  39.6
Short Mid           126     51  40.5
Rim                 115     68  59.1
Long Mid             67     27  40.3

--------------------------------

/tmp/ipython-input-662477314.py:218: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  quality_dist = df.dropna(subset=['shot_quality_tier']).groupby('shot_quality_tier').agg({




---



In [ ]:
"""
Las Vegas Aces - Get Team Roster from pbpstats
===============================================
Fetches team roster with player IDs for building lineup combinations.
NO API KEY REQUIRED

Based on: https://github.com/dblackrun/pbpstats-api-code-examples/blob/main/get-team-players-for-season.ipynb

Author: Krystal
Last Updated: 2026-01-06
"""

import requests
import pandas as pd
from pathlib import Path
import json

# ============================================================================
# CONFIGURATION
# ============================================================================

# Team configuration
TEAM_ID_WNBA = "1611661319"  # Las Vegas Aces pbpstats team ID
TEAM_ABBREV = "LV"
SEASONS = ["2024", "2025"]

# Data paths
DATA_DIR = Path("data")
RAW_DIR = DATA_DIR / "raw" / "pbpstats"
OUTPUT_DIR = DATA_DIR / "processed"

RAW_DIR.mkdir(parents=True, exist_ok=True)
OUTPUT_DIR.mkdir(parents=True, exist_ok=True)

# API endpoint
BASE_URL = "https://api.pbpstats.com"

print("="*70)
print("PBPSTATS - GET TEAM ROSTER")
print("="*70)
print(f"Team: {TEAM_ABBREV} (ID: {TEAM_ID_WNBA})")
print(f"Seasons: {', '.join(SEASONS)}")
print("="*70)
print()


# ============================================================================
# API FUNCTIONS
# ============================================================================

def get_team_players(season, team_id):
    """
    Get roster for a team in a season.
    Returns player IDs and names.

    Endpoint pattern from API examples:
    /get-team-players-for-season/{league}
    """
    print(f"\n[{season}] Fetching team roster...")

    url = f"{BASE_URL}/get-team-players-for-season/wnba"
    params = {
        "Season": season,
        "TeamId": team_id
    }

    try:
        response = requests.get(url, params=params, timeout=30)
        response.raise_for_status()

        data = response.json()

        # Save raw response
        raw_file = RAW_DIR / f"roster_{season}_raw.json"
        with open(raw_file, 'w') as f:
            json.dump(data, f, indent=2)
        print(f"  → Saved raw: {raw_file}")

        return data

    except requests.exceptions.RequestException as e:
        print(f"  ✗ API Error: {e}")
        return None


def parse_roster(data):
    """
    Parse roster response into clean dataframe.
    """
    if data is None:
        return pd.DataFrame()

    # The response structure varies by endpoint
    # Common patterns:
    # - data['players'] (list of player objects)
    # - data (directly as list)
    # - data['results']

    if isinstance(data, list):
        players = data
    elif 'players' in data:
        players = data['players']
    elif 'results' in data:
        players = data['results']
    else:
        print(f"  ⚠ Unknown response structure: {list(data.keys())}")
        return pd.DataFrame()

    if not players:
        return pd.DataFrame()

    df = pd.DataFrame(players)

    # Standardize column names
    col_mapping = {
        'PlayerId': 'player_id',
        'player_id': 'player_id',
        'id': 'player_id',
        'PlayerName': 'player_name',
        'player_name': 'player_name',
        'name': 'player_name',
        'Name': 'player_name'
    }

    df = df.rename(columns=col_mapping)

    return df


# ============================================================================
# LINEUP COMBINATION BUILDER
# ============================================================================

def build_lineup_ids(player_ids, lineup_size=5):
    """
    Build all possible lineup combinations.
    Lineup IDs are hyphen-separated player IDs sorted as strings.

    For Top 8 rotation, this generates 56 possible 5-man lineups.
    """
    from itertools import combinations

    # Sort player IDs as strings (pbpstats requirement)
    player_ids = sorted([str(pid) for pid in player_ids])

    lineups = []
    for combo in combinations(player_ids, lineup_size):
        # Sort the combination and join with hyphens
        lineup_id = '-'.join(sorted(combo))
        lineups.append(lineup_id)

    return lineups


def identify_top_rotation(wehoop_player_ids_csv=None, top_n=8):
    """
    Identify top rotation players from wehoop data.

    Args:
        wehoop_player_ids_csv: Path to wehoop player IDs file
        top_n: Number of players to include (default 8)
    """
    if wehoop_player_ids_csv is None:
        print("\n⚠ No wehoop player IDs file provided")
        print("  You'll need to manually select top rotation players")
        return None

    try:
        df = pd.read_csv(wehoop_player_ids_csv)

        # Assume it has columns: player_id, player_name, total_minutes
        # Take top N by minutes
        top_players = df.nlargest(top_n, 'total_minutes')

        print(f"\n[TOP {top_n}] Identified from wehoop data:")
        print(top_players[['player_name', 'total_minutes']].to_string(index=False))

        return top_players['player_id'].tolist()

    except Exception as e:
        print(f"  ✗ Error reading wehoop data: {e}")
        return None


# ============================================================================
# MAIN EXECUTION
# ============================================================================

def main():
    print(f"\nStarted: {pd.Timestamp.now()}\n")

    all_rosters = []

    for season in SEASONS:
        print("="*70)
        print(f"PROCESSING SEASON {season}")
        print("="*70)

        # Fetch roster
        data = get_team_players(season, TEAM_ID_WNBA)

        if data:
            # Parse
            roster = parse_roster(data)

            if len(roster) > 0:
                roster['season'] = season
                all_rosters.append(roster)

                print(f"  ✓ Found {len(roster)} players")

                # Show sample
                if 'player_name' in roster.columns and 'player_id' in roster.columns:
                    print("\n  Sample roster:")
                    print(roster[['player_name', 'player_id']].head(10).to_string(index=False))

    # Combine seasons
    if all_rosters:
        print("\n" + "="*70)
        print("COMBINING SEASONS")
        print("="*70)

        roster_combined = pd.concat(all_rosters, ignore_index=True)

        # Get unique players across seasons
        unique_players = roster_combined.drop_duplicates(subset=['player_id'])

        print(f"\n✓ Total unique players: {len(unique_players)}")

        # Save
        output_file = OUTPUT_DIR / "aces_roster_pbpstats.csv"
        unique_players.to_csv(output_file, index=False)
        print(f"✓ Saved: {output_file}")

        # Try to merge with wehoop player IDs
        wehoop_file = OUTPUT_DIR / "aces_player_ids.csv"
        if wehoop_file.exists():
            print("\n" + "="*70)
            print("MERGING WITH WEHOOP DATA")
            print("="*70)

            wehoop_players = pd.read_csv(wehoop_file)

            # Merge to get minutes data
            # Note: player IDs might not match between wehoop and pbpstats
            # wehoop uses ESPN IDs, pbpstats might use different IDs

            merged = unique_players.merge(
                wehoop_players,
                on='player_id',
                how='outer',
                suffixes=('_pbp', '_wehoop')
            )

            print(f"  ✓ Merged {len(merged)} players")

            # Save merged
            merged_file = OUTPUT_DIR / "aces_players_merged.csv"
            merged.to_csv(merged_file, index=False)
            print(f"  ✓ Saved: {merged_file}")

            # Identify top rotation
            if 'total_minutes' in merged.columns:
                top_players = identify_top_rotation(wehoop_file, top_n=8)

                if top_players:
                    # Build lineup combinations
                    print("\n" + "="*70)
                    print("BUILDING LINEUP COMBINATIONS")
                    print("="*70)

                    lineups = build_lineup_ids(top_players, lineup_size=5)

                    print(f"  ✓ Generated {len(lineups)} possible 5-man lineups")
                    print(f"\n  Sample lineup IDs:")
                    for lineup_id in lineups[:5]:
                        print(f"    {lineup_id}")

                    # Save lineup IDs
                    lineup_df = pd.DataFrame({'lineup_id': lineups})
                    lineup_file = OUTPUT_DIR / "aces_lineup_ids.csv"
                    lineup_df.to_csv(lineup_file, index=False)
                    print(f"\n  ✓ Saved: {lineup_file}")

    print("\n" + "="*70)
    print("NEXT STEPS")
    print("="*70)
    print("""
1. Check output files:
   - aces_roster_pbpstats.csv (roster from pbpstats)
   - aces_players_merged.csv (pbpstats + wehoop merged)
   - aces_lineup_ids.csv (Top 8 lineup combinations)

2. Use lineup IDs with pbpstats lineup script:
   For each lineup_id in aces_lineup_ids.csv:
   - Call get-lineup-opponent-summary endpoint
   - Get lineup stats vs opponent lineups

3. If player IDs don't match between sources:
   - You may need to manually map player names
   - Or use only wehoop player IDs
   - pbpstats player IDs come from stats.nba.com
    """)
    print("="*70)

    print(f"\nCompleted: {pd.Timestamp.now()}")


if __name__ == "__main__":
    main()

PBPSTATS - GET TEAM ROSTER
Team: LV (ID: 1611661319)
Seasons: 2024, 2025


Started: 2026-01-07 06:32:31.429597

PROCESSING SEASON 2024

[2024] Fetching team roster...
  ✗ API Error: 404 Client Error: Not Found for url: https://api.pbpstats.com/get-team-players-for-season/wnba?Season=2024&TeamId=1611661319
PROCESSING SEASON 2025

[2025] Fetching team roster...
  ✗ API Error: 404 Client Error: Not Found for url: https://api.pbpstats.com/get-team-players-for-season/wnba?Season=2025&TeamId=1611661319

NEXT STEPS

1. Check output files:
   - aces_roster_pbpstats.csv (roster from pbpstats)
   - aces_players_merged.csv (pbpstats + wehoop merged)
   - aces_lineup_ids.csv (Top 8 lineup combinations)

2. Use lineup IDs with pbpstats lineup script:
   For each lineup_id in aces_lineup_ids.csv:
   - Call get-lineup-opponent-summary endpoint
   - Get lineup stats vs opponent lineups

3. If player IDs don't match between sources:
   - You may need to manually map player names
   - Or use only wehoop

In [ ]:
players_url = "https://api.pbpstats.com/get-team-players-for-season"
players_params = {
    "Season": "2024",
    "SeasonType": "Regular Season",
    "TeamId": "1611661319"
}
players_response = requests.get(players_url, params=players_params)
players = players_response.json()
players["players"]

{'1642305': 'Jessika Carter',
 '1631031': 'Queen Egbo',
 '1628932': "A'ja Wilson",
 '1628276': 'Kelsey Plum',
 '204329': 'Kiah Stokes',
 '1631006': 'Kierstan Bell',
 '202641': 'Sydney Colson',
 '203833': 'Chelsea Gray',
 '1629498': 'Jackie Young',
 '1629484': 'Megan Gustafson',
 '1642324': 'Kate Martin',
 '1642300': 'Dyaisha Fair',
 '202252': 'Alysha Clark',
 '1628242': 'Emma Cannon',
 '203026': 'Tiffany Hayes'}

In [ ]:
players_url = "https://api.pbpstats.com/get-team-players-for-season"
players_params = {
    "Season": "2025",
    "SeasonType": "Regular Season",
    "TeamId": "1611661319"
}
players_response = requests.get(players_url, params=players_params)
players = players_response.json()
players["players"]

{'1629498': 'Jackie Young',
 '204319': 'Jewell Loyd',
 '1629484': 'Megan Gustafson',
 '1630115': 'Joyner Holmes',
 '203833': 'Chelsea Gray',
 '1631019': 'NaLyssa Smith',
 '204323': 'Cheyenne Parker-Tyus',
 '1628932': "A'ja Wilson",
 '204329': 'Kiah Stokes',
 '1627676': 'Tiffany Mitchell',
 '1642801': 'Aaliyah Nye',
 '1631006': 'Kierstan Bell',
 '1630389': 'Dana Evans',
 '1642295': 'Elizabeth Kitley',
 '204325': 'Crystal Bradford'}

In [ ]:
"""
Las Vegas Aces - pbpstats Lineup Data Pull (Python)
====================================================
Pulls lineup combinations and opponent matchups from pbpstats.com
NO API KEY REQUIRED - free public API

Author: Krystal
Last Updated: 2026-01-06
"""

import requests
import pandas as pd
import json
from pathlib import Path
from itertools import combinations
import time

# ============================================================================
# CONFIGURATION
# ============================================================================

# Team configuration
TEAM_ID_WNBA = "1611661319"  # Las Vegas Aces pbpstats team ID
TEAM_ABBREV = "LV"
SEASONS = ["2024"]  # WNBA uses single year format (not "2024-25")

# Data paths
DATA_DIR = Path("data")
RAW_DIR = DATA_DIR / "raw" / "pbpstats"
OUTPUT_DIR = DATA_DIR / "master"

RAW_DIR.mkdir(parents=True, exist_ok=True)
OUTPUT_DIR.mkdir(parents=True, exist_ok=True)

# API configuration
BASE_URL = "https://api.pbpstats.com"
REQUEST_DELAY = 0.5  # Seconds between API calls (be respectful)

print("="*70)
print("PBPSTATS LINEUP DATA PULL (PYTHON)")
print("="*70)
print(f"Team: {TEAM_ABBREV} (ID: {TEAM_ID_WNBA})")
print(f"Seasons: {', '.join(SEASONS)}")
print(f"Data: Lineup combinations + opponent matchups")
print("="*70)
print()


# ============================================================================
# API FUNCTIONS
# ============================================================================

def make_api_request(endpoint, params):
    """Make GET request to pbpstats API. No API key required!"""
    url = f"{BASE_URL}{endpoint}"

    try:
        response = requests.get(url, params=params, timeout=30)
        response.raise_for_status()
        return response.json()
    except requests.exceptions.RequestException as e:
        print(f"  ✗ API Error: {e}")
        return None


def get_lineup_opponent_summary(season, team_id, lineup_id):
    """
    Get lineup stats vs each opponent lineup.

    Returns:
    - Minutes, possessions played together
    - ORtg, DRtg, NetRtg vs each opponent lineup
    - Four Factors for lineup and opponent
    - Plus/minus data

    Args:
        season: "2024" or "2025" (WNBA uses single year)
        team_id: Team ID from pbpstats
        lineup_id: Hyphen-separated player IDs (sorted as strings)
                   Example: "1628276-1628932-1629498-203833-204324"
    """
    print(f"\n[LINEUP] Fetching: {lineup_id[:30]}...")

    endpoint = "/get-lineup-opponent-summary/wnba"
    params = {
        "LineupId": lineup_id,
        "TeamId": team_id,
        "Season": season,
        "SeasonType": "Regular Season"
    }

    data = make_api_request(endpoint, params)

    if data is None:
        print(f"  ✗ Failed")
        return pd.DataFrame()

    # Save raw response
    safe_lineup_id = lineup_id.replace('-', '_')
    raw_file = RAW_DIR / f"lineup_{safe_lineup_id}_{season}_raw.json"
    with open(raw_file, 'w') as f:
        json.dump(data, f, indent=2)
    print(f"  → Saved raw: {raw_file.name}")

    # Parse results
    results = data.get('results', [])

    if not results:
        print(f"  ⚠ No opponent matchups")
        return pd.DataFrame()

    # Convert to dataframe
    df = pd.DataFrame(results)

    # Add metadata
    df['our_lineup_id'] = lineup_id
    df['season'] = season
    df['team_id'] = team_id

    print(f"  ✓ Found {len(df)} opponent lineups")

    return df


# ============================================================================
# LINEUP COMBINATION BUILDER
# ============================================================================

def build_lineup_ids_from_csv(player_ids_csv, top_n=8):
    """
    Build lineup combinations from wehoop player IDs.

    Args:
        player_ids_csv: Path to CSV with player_id column
        top_n: Number of top players to use (8 = 56 lineups, 10 = 252 lineups)

    Returns:
        List of lineup IDs (hyphen-separated, sorted)
    """
    print("\n" + "="*70)
    print(f"BUILDING LINEUP IDS FROM TOP {top_n} ROTATION")
    print("="*70)

    try:
        df = pd.read_csv(player_ids_csv)

        # Get top N by minutes
        if 'total_minutes' in df.columns:
            top_players = df.nlargest(top_n, 'total_minutes')
        else:
            print("  ⚠ No total_minutes column, using first {top_n} players")
            top_players = df.head(top_n)

        print(f"\n  Top {top_n} players:")
        if 'player_name' in df.columns and 'total_minutes' in df.columns:
            print(top_players[['player_name', 'total_minutes']].to_string(index=False))

        # Get player IDs
        player_ids = top_players['player_id'].astype(str).tolist()

        # Generate all 5-player combinations
        lineups = []
        for combo in combinations(player_ids, 5):
            # Sort and join with hyphens (pbpstats requirement)
            lineup_id = '-'.join(sorted(combo))
            lineups.append(lineup_id)

        print(f"\n  ✓ Generated {len(lineups)} possible 5-man lineups")
        print(f"\n  Sample lineup IDs:")
        for lineup_id in lineups[:3]:
            print(f"    {lineup_id}")

        return lineups

    except Exception as e:
        print(f"  ✗ Error: {e}")
        return []


def manual_lineup_ids():
    """
    If you know specific lineup IDs, enter them here.
    Example: 2022 Aces starting lineup
    """
    return [
        # Chelsea Gray, Dearica Hamby, Kelsey Plum, A'ja Wilson, Jackie Young
        "1628276-1628932-1629498-203833-204324"
    ]


# ============================================================================
# MAIN EXECUTION
# ============================================================================

def main():
    print(f"\nStarted: {pd.Timestamp.now()}\n")

    # Option 1: Build from wehoop player IDs
    player_ids_file = DATA_DIR / "processed" / "aces_player_ids.csv"

    if player_ids_file.exists():
        print(f"✓ Found wehoop player IDs: {player_ids_file}")
        lineup_ids = build_lineup_ids_from_csv(player_ids_file, top_n=8)
    else:
        print(f"⚠ No wehoop player IDs found at: {player_ids_file}")
        print(f"  Using manual lineup IDs instead...")
        lineup_ids = manual_lineup_ids()

    if not lineup_ids:
        print("\n✗ No lineup IDs to process!")
        print("\nTo generate lineup IDs:")
        print("  1. Run calculate_metrics_from_csv.py first")
        print("  2. It will create data/processed/aces_player_ids.csv")
        print("  3. Then run this script again")
        return

    # Fetch lineup data
    all_lineups = []

    for season in SEASONS:
        print("\n" + "="*70)
        print(f"PROCESSING SEASON {season}")
        print("="*70)
        print(f"Fetching {len(lineup_ids)} lineups...")

        for i, lineup_id in enumerate(lineup_ids, 1):
            print(f"\n[{i}/{len(lineup_ids)}]", end=" ")

            # Fetch lineup opponent summary
            df = get_lineup_opponent_summary(season, TEAM_ID_WNBA, lineup_id)

            if len(df) > 0:
                all_lineups.append(df)

            # Rate limiting - be respectful to free API
            if i < len(lineup_ids):
                time.sleep(REQUEST_DELAY)

    # Combine all data
    if all_lineups:
        print("\n" + "="*70)
        print("COMBINING DATA")
        print("="*70)

        lineups_combined = pd.concat(all_lineups, ignore_index=True)

        print(f"\n✓ Total matchups: {len(lineups_combined)}")
        print(f"  Unique lineups: {lineups_combined['our_lineup_id'].nunique()}")
        print(f"  Unique opponents: {lineups_combined['lineup_id'].nunique() if 'lineup_id' in lineups_combined.columns else 'N/A'}")

        # Save to master
        output_file = OUTPUT_DIR / "aces_lineups_pbpstats.parquet"
        lineups_combined.to_parquet(output_file, index=False)
        print(f"\n✓ Saved: {output_file}")

        # Also save as CSV for easy viewing
        csv_file = OUTPUT_DIR / "aces_lineups_pbpstats.csv"
        lineups_combined.to_csv(csv_file, index=False)
        print(f"✓ Saved: {csv_file}")

        # Show summary stats
        print("\n" + "="*70)
        print("LINEUP SUMMARY")
        print("="*70)

        # Top lineups by minutes
        lineup_summary = lineups_combined.groupby('our_lineup_id').agg({
            'minutes': 'sum',
            'total_possessions': 'sum',
            'ortg': 'first',
            'drtg': 'first',
            'net_rtg': 'first'
        }).sort_values('minutes', ascending=False)

        print("\nTop 5 lineups by minutes:")
        print(lineup_summary.head().to_string())

    else:
        print("\n⚠ No lineup data fetched")

    print("\n" + "="*70)
    print("NEXT STEPS")
    print("="*70)
    print("""
1. Load into Tableau:
   - Open aces_lineups_pbpstats.parquet or .csv
   - Join with aces_player_box_metrics.csv to get player names

2. Analysis ideas:
   - Which lineups have highest NetRtg?
   - Which lineups struggle vs certain opponent lineups?
   - Compare lineup efficiency in different situations

3. Merge with wehoop data:
   - Add game context (home/away, opponent strength)
   - Combine lineup stats with individual player stats
   - Create comprehensive lineup profiles

AVAILABLE COLUMNS:
- our_lineup_id: Your lineup ID
- lineup_id: Opponent lineup ID
- minutes, seconds_played: Time together
- total_possessions: Possession count
- ortg, drtg, net_rtg: Efficiency ratings
- efg, tov_pct, ftr: Your Four Factors
- opp_efg, opp_tov_pct, opp_ftr: Opponent Four Factors
- pts_for, pts_against, plus_minus: Scoring
- fg2a, fg2m, fg3a, fg3m: Shooting breakdown
    """)
    print("="*70)

    print(f"\nCompleted: {pd.Timestamp.now()}")


if __name__ == "__main__":
    main()

PBPSTATS LINEUP DATA PULL (PYTHON)
Team: LV (ID: 1611661319)
Seasons: 2024
Data: Lineup combinations + opponent matchups


Started: 2026-01-07 06:36:47.466386

⚠ No wehoop player IDs found at: data/processed/aces_player_ids.csv
  Using manual lineup IDs instead...

PROCESSING SEASON 2024
Fetching 1 lineups...

[1/1] 
[LINEUP] Fetching: 1628276-1628932-1629498-203833...
  → Saved raw: lineup_1628276_1628932_1629498_203833_204324_2024_raw.json
  ⚠ No opponent matchups

⚠ No lineup data fetched

NEXT STEPS

1. Load into Tableau:
   - Open aces_lineups_pbpstats.parquet or .csv
   - Join with aces_player_box_metrics.csv to get player names

2. Analysis ideas:
   - Which lineups have highest NetRtg?
   - Which lineups struggle vs certain opponent lineups?
   - Compare lineup efficiency in different situations

3. Merge with wehoop data:
   - Add game context (home/away, opponent strength)
   - Combine lineup stats with individual player stats
   - Create comprehensive lineup profiles

AVAILA

In [ ]:
url = "https://api.pbpstats.com/get-lineup-subunit-stats/wnba"
params = {
    "LineupId": "1628276-1628932-1629498-203833-204324", # lineup ids are hyphen separated player ids sorted as strings
    "TeamId": "1611661319",
    "Season": "2024",
    "SeasonType": "Regular Season",
    "SubUnitSize": 2
}
response = requests.get(url, params=params)
response_json = response.json()
response_json['results'][0]

{'OffRtg': 107.15237672145713,
 'DefRtg': 100.17841213202497,
 'NetRtg': 6.973964589432157,
 'Minutes': 1114.0,
 'On': "Kelsey Plum, A'ja Wilson",
 'Off': '',
 'TeamWowyQueryParams': {'Season': '2024',
  'SeasonType': 'Regular Season',
  'TeamId': '1611661319',
  'Type': 'Team',
  '0Exactly2OnFloor': '1628276,1628932'},
 'OpponentWowyQueryParams': {'Season': '2024',
  'SeasonType': 'Regular Season',
  'TeamId': '1611661319',
  'Type': 'Opponent',
  '0Exactly2OnFloor': '1628276,1628932'},
 'RowId': 'On:1628276,1628932',
 'Fg3Pct': 0.3452722063037249,
 'Fg2Pct': 0.5181518151815182,
 'OppFg3Pct': 0.3475409836065574,
 'OppFg2Pct': 0.4630177514792899}

'1642305': 'Jessika Carter',
 '1631031': 'Queen Egbo',
 '1628932': "A'ja Wilson",
 '1628276': 'Kelsey Plum',
 '204329': 'Kiah Stokes',
 '1631006': 'Kierstan Bell',
 '202641': 'Sydney Colson',
 '203833': 'Chelsea Gray',
 '1629498': 'Jackie Young',
 '1629484': 'Megan Gustafson',
 '1642324': 'Kate Martin',
 '1642300': 'Dyaisha Fair',
 '202252': 'Alysha Clark',
 '1628242': 'Emma Cannon',
 '203026': 'Tiffany Hayes

In [ ]:
import requests
import pandas as pd
import json
from pathlib import Path
from itertools import combinations
import time

url = "https://api.pbpstats.com/get-on-off/wnba/player" # Changed from 'nba' to 'wnba'
params = {
    "Season": "2024",
    "SeasonType": "Regular Season",
    "TeamId": "1611661319",
    "PlayerId": "1628932"
    }
response = requests.get(url, params=params)
response_json = response.json()

# Check if 'results' key exists before accessing it
if 'results' in response_json:
    print(response_json['results'])
else:
    print(f"API response did not contain 'results' key. Response: {response_json}")

{'Assisted2sPct': [{'Name': 'Alysha Clark', 'On': '0.880', 'Off': '0.833', 'On-Off': '0.047', 'MinutesOn': 749, 'MinutesOff': 188}, {'Name': 'Sydney Colson', 'On': '0.500', 'Off': '0.167', 'On-Off': '0.333', 'MinutesOn': 114, 'MinutesOff': 116}, {'Name': 'Tiffany Hayes', 'On': '0.534', 'Off': '0.529', 'On-Off': '0.005', 'MinutesOn': 538, 'MinutesOff': 147}, {'Name': 'Chelsea Gray', 'On': '0.293', 'Off': '0.294', 'On-Off': '-0.001', 'MinutesOn': 564, 'MinutesOff': 137}, {'Name': 'Kiah Stokes', 'On': '1.000', 'Off': '1.000', 'On-Off': '0.000', 'MinutesOn': 509, 'MinutesOff': 195}, {'Name': 'Emma Cannon', 'On': '0.000', 'Off': '1.000', 'On-Off': '-1.000', 'MinutesOn': 1, 'MinutesOff': 8}, {'Name': 'Kelsey Plum', 'On': '0.386', 'Off': '0.333', 'On-Off': '0.053', 'MinutesOn': 1114, 'MinutesOff': 145}, {'Name': 'Megan Gustafson', 'On': '0.889', 'Off': '0.773', 'On-Off': '0.116', 'MinutesOn': 233, 'MinutesOff': 168}, {'Name': 'Jackie Young', 'On': '0.495', 'Off': '0.444', 'On-Off': '0.051', '

In [ ]:
import pandas as pd

all_player_on_off_stats = []

# response_json in the kernel state has the structure {'results': {...}}
# So we need to access response_json['results']
if 'results' in response_json:
    player_on_off_data = response_json['results']
    for metric_name, stats_list in player_on_off_data.items():
        df_metric = pd.DataFrame(stats_list)
        df_metric['Metric'] = metric_name
        all_player_on_off_stats.append(df_metric)

    if all_player_on_off_stats:
        combined_player_on_off_stats = pd.concat(all_player_on_off_stats, ignore_index=True)
        print("Combined Player On-Off Stats DataFrame created.")
        display(combined_player_on_off_stats.head())

        # Check for requested fields
        print("\nChecking for requested fields:")
        if 'Name' in combined_player_on_off_stats.columns:
            print("  'Name' column found (player name).")
        if 'player_id' in combined_player_on_off_stats.columns:
            print("  'player_id' column found.")
        if 'player_name' in combined_player_on_off_stats.columns:
            print("  'player_name' column found.")
        if 'Season' in combined_player_on_off_stats.columns:
            print("  'Season' column found.")
    else:
        print("No player on-off stats were found to combine.")
else:
    print("API response did not contain 'results' key, so no player on-off data available.")


Combined Player On-Off Stats DataFrame created.


Name     On    Off  On-Off  MinutesOn  MinutesOff         Metric
0   Alysha Clark  0.880  0.833   0.047        749         188  Assisted2sPct
1  Sydney Colson  0.500  0.167   0.333        114         116  Assisted2sPct
2  Tiffany Hayes  0.534  0.529   0.005        538         147  Assisted2sPct
3   Chelsea Gray  0.293  0.294  -0.001        564         137  Assisted2sPct
4    Kiah Stokes  1.000  1.000   0.000        509         195  Assisted2sPct


Checking for requested fields:
  'Name' column found (player name).


In [ ]:
# Display unique metric names to identify the correct one for NetRtg
print("Unique metrics available in the DataFrame:")
print(combined_player_on_off_stats['Metric'].unique())

# The code will stop here for you to review the available metrics.
# Once you identify the correct metric for Net Rating, please provide it.
# For example, if it's 'NetRating', you would then use:
# net_rtg_impact = combined_player_on_off_stats[
#     combined_player_on_off_stats['Metric'] == 'NetRating'
# ].copy()

Unique metrics available in the DataFrame:
['Assisted2sPct' 'NonPutbacksAssisted2sPct' 'Assisted3sPct' 'Fg3Pct'
 'NonHeaveFg3Pct' 'Fg2Pct' 'EfgPct' 'TsPct' 'FG3APct' 'FG3APctBlocked'
 'FG2APctBlocked' 'Usage' 'LiveBallTurnoverPct' 'DefFTReboundPct'
 'OffFTReboundPct' 'DefTwoPtReboundPct' 'OffTwoPtReboundPct'
 'DefThreePtReboundPct' 'OffThreePtReboundPct' 'DefFGReboundPct'
 'OffFGReboundPct' 'OffAtRimReboundPct' 'OffShortMidRangeReboundPct'
 'OffLongMidRangeReboundPct' 'OffArc3ReboundPct' 'OffCorner3ReboundPct'
 'DefAtRimReboundPct' 'DefShortMidRangeReboundPct'
 'DefLongMidRangeReboundPct' 'DefArc3ReboundPct' 'DefCorner3ReboundPct'
 'BlocksRecoveredPct' 'SecondsPerPossOff' 'SecondsPerPossDef'
 'AtRimFrequency' 'AtRimAccuracy' 'AtRimPctAssisted'
 'ShortMidRangeFrequency' 'ShortMidRangeAccuracy'
 'ShortMidRangePctAssisted' 'LongMidRangeFrequency' 'LongMidRangeAccuracy'
 'LongMidRangePctAssisted' 'Corner3Frequency' 'Corner3Accuracy'
 'Corner3PctAssisted' 'Arc3Frequency' 'Arc3Accuracy' 'Arc



---


In [1]:
"""
Las Vegas Aces - Identify Top 8 Rotation (2025)
================================================
Filters raw wehoop parquet files to Las Vegas Aces only.
Identifies Top 8 players by total minutes for player profiles.

Author: Krystal
Last Updated: 2026-01-07
"""

import pandas as pd
from pathlib import Path

# ============================================================================
# CONFIGURATION
# ============================================================================

# Input paths - UPDATE THESE to your actual parquet file locations
PLAYER_BOX_PARQUET = "/content/player_box_2025.parquet"  # Your raw wehoop player box scores
TEAM_BOX_PARQUET = "/content/team_box_2025.parquet"  # Your raw wehoop team box scores (optional)

# Output paths
OUTPUT_DIR = Path("data/processed")
OUTPUT_DIR.mkdir(parents=True, exist_ok=True)

# Team filter - Las Vegas Aces
# wehoop uses different team identifiers - adjust as needed
LVA_TEAM_IDS = ["16"]  # wehoop team_id for Las Vegas Aces
LVA_TEAM_NAMES = ["Las Vegas Aces", "LV Aces", "LVA", "Las Vegas"]

print("="*70)
print("LAS VEGAS ACES - TOP 8 ROTATION IDENTIFIER")
print("="*70)
print(f"Input: {PLAYER_BOX_PARQUET}")
print(f"Team: Las Vegas Aces")
print(f"Season: 2025")
print("="*70)
print()


# ============================================================================
# LOAD AND FILTER DATA
# ============================================================================

print("[STEP 1] Loading raw player box scores...")

try:
    # Load parquet file
    df = pd.read_parquet(PLAYER_BOX_PARQUET)
    print(f"✓ Loaded {len(df):,} rows")
    print(f"  Columns: {list(df.columns[:15])}...")

    # Show unique teams to identify correct filter
    print(f"\n  Checking team identifiers in data...")
    if 'team_id' in df.columns:
        unique_teams = df['team_id'].unique()
        print(f"  Unique team_ids: {sorted(unique_teams)}")

    if 'team_display_name' in df.columns:
        unique_team_names = df['team_display_name'].unique()
        print(f"  Unique team names: {sorted(unique_team_names)[:10]}...")

except FileNotFoundError:
    print(f"✗ Could not find: {PLAYER_BOX_PARQUET}")
    print("  Please update PLAYER_BOX_PARQUET path in script")
    exit(1)


# ============================================================================
# FILTER TO LAS VEGAS ACES
# ============================================================================

print("\n[STEP 2] Filtering to Las Vegas Aces...")

# Try multiple filter approaches
df_lva = pd.DataFrame()

# Approach 1: Filter by team_id
if 'team_id' in df.columns:
    df_lva = df[df['team_id'].isin(LVA_TEAM_IDS)]
    print(f"  → By team_id: {len(df_lva):,} rows")

# Approach 2: Filter by team name (if team_id didn't work)
if len(df_lva) == 0 and 'team_display_name' in df.columns:
    df_lva = df[df['team_display_name'].isin(LVA_TEAM_NAMES)]
    print(f"  → By team_display_name: {len(df_lva):,} rows")

# Approach 3: Try partial match on team name
if len(df_lva) == 0 and 'team_display_name' in df.columns:
    df_lva = df[df['team_display_name'].str.contains('Vegas|Aces|LV', case=False, na=False)]
    print(f"  → By partial match: {len(df_lva):,} rows")

# Approach 4: Try abbreviation column
if len(df_lva) == 0 and 'team_abbreviation' in df.columns:
    df_lva = df[df['team_abbreviation'].isin(['LV', 'LVA'])]
    print(f"  → By team_abbreviation: {len(df_lva):,} rows")

if len(df_lva) == 0:
    print(f"\n✗ Could not filter to Las Vegas Aces!")
    print(f"  Please check team identifier columns and update LVA_TEAM_IDS or LVA_TEAM_NAMES")
    print(f"\n  Available team columns:")
    team_cols = [col for col in df.columns if 'team' in col.lower()]
    print(f"  {team_cols}")
    exit(1)

print(f"\n✓ Filtered to {len(df_lva):,} Las Vegas Aces player-game rows")

# Show sample
if 'athlete_display_name' in df_lva.columns:
    print(f"\n  Sample players:")
    sample_players = df_lva['athlete_display_name'].value_counts().head(5)
    print(sample_players)


# ============================================================================
# SAVE FILTERED DATA
# ============================================================================

print("\n[STEP 3] Saving filtered data...")

output_file = OUTPUT_DIR / "aces_player_box_2025.csv"
df_lva.to_csv(output_file, index=False)
print(f"✓ Saved: {output_file}")
print(f"  Rows: {len(df_lva):,}")
print(f"  Columns: {len(df_lva.columns)}")


# ============================================================================
# IDENTIFY TOP 8 ROTATION
# ============================================================================

print("\n[STEP 4] Identifying Top 8 rotation players...")

# Determine player identifier columns
player_id_col = None
player_name_col = None

for col in ['athlete_id', 'player_id', 'id']:
    if col in df_lva.columns:
        player_id_col = col
        break

for col in ['athlete_display_name', 'player_name', 'name']:
    if col in df_lva.columns:
        player_name_col = col
        break

if not player_id_col or not player_name_col:
    print(f"✗ Could not find player ID or name columns")
    print(f"  Available columns: {list(df_lva.columns)}")
    exit(1)

print(f"  Using: {player_id_col} and {player_name_col}")

# Determine minutes column
minutes_col = None
for col in ['minutes', 'min', 'mp']:
    if col in df_lva.columns:
        minutes_col = col
        break

if not minutes_col:
    print(f"✗ Could not find minutes column")
    exit(1)

# Convert minutes to numeric (might be "MM:SS" format)
if df_lva[minutes_col].dtype == 'object':
    print(f"  Converting minutes from MM:SS format...")
    try:
        df_lva['minutes_decimal'] = df_lva[minutes_col].apply(
            lambda x: int(str(x).split(':')[0]) + int(str(x).split(':')[1])/60
            if isinstance(x, str) and ':' in str(x)
            else float(x) if pd.notna(x) else 0
        )
    except:
        df_lva['minutes_decimal'] = pd.to_numeric(df_lva[minutes_col], errors='coerce').fillna(0)
else:
    df_lva['minutes_decimal'] = pd.to_numeric(df_lva[minutes_col], errors='coerce').fillna(0)

# Aggregate by player
player_summary = df_lva.groupby([player_id_col, player_name_col]).agg({
    'minutes_decimal': 'sum',
    'game_id': 'nunique'  # Games played
}).reset_index()

player_summary.columns = ['player_id', 'player_name', 'total_minutes', 'games']

# Calculate minutes per game
player_summary['mpg'] = player_summary['total_minutes'] / player_summary['games']

# Sort by total minutes
player_summary = player_summary.sort_values('total_minutes', ascending=False)

# Get Top 8
top_8 = player_summary.head(8).copy()

print(f"\n✓ Top 8 rotation by total minutes:")
print("="*70)
print(top_8.to_string(index=False))
print("="*70)

# Save Top 8 list
top_8_file = OUTPUT_DIR / "aces_top8_rotation_2025.csv"
top_8.to_csv(top_8_file, index=False)
print(f"\n✓ Saved Top 8 list: {top_8_file}")

# Also save full player summary
summary_file = OUTPUT_DIR / "aces_player_summary_2025.csv"
player_summary.to_csv(summary_file, index=False)
print(f"✓ Saved full player summary: {summary_file}")


# ============================================================================
# ADDITIONAL STATS FOR TOP 8
# ============================================================================

print("\n[STEP 5] Calculating detailed stats for Top 8...")

# Get detailed stats for top 8 players
top_8_ids = top_8['player_id'].tolist()
df_top8 = df_lva[df_lva[player_id_col].isin(top_8_ids)].copy()

# Determine available stat columns
stat_cols = []
for col in ['field_goals_made', 'fgm', 'field_goals_attempted', 'fga',
            'three_point_field_goals_made', 'fg3m', 'three_point_field_goals_attempted', 'fg3a',
            'free_throws_made', 'ftm', 'free_throws_attempted', 'fta',
            'points', 'pts', 'assists', 'ast', 'rebounds', 'reb',
            'offensive_rebounds', 'orb', 'defensive_rebounds', 'drb',
            'steals', 'stl', 'blocks', 'blk', 'turnovers', 'tov']:
    if col in df_top8.columns:
        stat_cols.append(col)

print(f"  Found {len(stat_cols)} stat columns")

# Aggregate detailed stats
agg_dict = {'minutes_decimal': 'sum', 'game_id': 'nunique'}
for col in stat_cols:
    if col in df_top8.columns:
        agg_dict[col] = 'sum'

detailed_stats = df_top8.groupby([player_id_col, player_name_col]).agg(agg_dict).reset_index()

# Calculate per-game averages
for col in stat_cols:
    if col in detailed_stats.columns and col not in ['game_id', 'minutes_decimal']:
        detailed_stats[f'{col}_pg'] = detailed_stats[col] / detailed_stats['game_id']

# Calculate shooting percentages
if 'field_goals_made' in detailed_stats.columns and 'field_goals_attempted' in detailed_stats.columns:
    detailed_stats['fg_pct'] = detailed_stats['field_goals_made'] / detailed_stats['field_goals_attempted']
elif 'fgm' in detailed_stats.columns and 'fga' in detailed_stats.columns:
    detailed_stats['fg_pct'] = detailed_stats['fgm'] / detailed_stats['fga']

if 'three_point_field_goals_made' in detailed_stats.columns and 'three_point_field_goals_attempted' in detailed_stats.columns:
    detailed_stats['fg3_pct'] = detailed_stats['three_point_field_goals_made'] / detailed_stats['three_point_field_goals_attempted']
elif 'fg3m' in detailed_stats.columns and 'fg3a' in detailed_stats.columns:
    detailed_stats['fg3_pct'] = detailed_stats['fg3m'] / detailed_stats['fg3a']

# Sort by minutes
detailed_stats = detailed_stats.sort_values('minutes_decimal', ascending=False)

# Save
detailed_file = OUTPUT_DIR / "aces_top8_detailed_stats_2025.csv"
detailed_stats.to_csv(detailed_file, index=False)
print(f"✓ Saved detailed stats: {detailed_file}")

# Show summary
print(f"\nTop 8 Summary:")
display_cols = [player_name_col, 'game_id', 'minutes_decimal']
if 'points' in detailed_stats.columns:
    display_cols.append('points')
if 'pts_pg' in detailed_stats.columns:
    display_cols.append('pts_pg')
if 'fg_pct' in detailed_stats.columns:
    display_cols.append('fg_pct')

print(detailed_stats[display_cols].head(8).to_string(index=False))


# ============================================================================
# NEXT STEPS
# ============================================================================

print("\n" + "="*70)
print("NEXT STEPS")
print("="*70)
print("""
1. Manual On/Off Data Collection (10 min):
   For each of the Top 8 players, visit:
   https://www.pbpstats.com/on-off/wnba/player?Season=2024-25&SeasonType=Regular+Season&TeamId=1611661319&PlayerId=XXXXXX

   Record:
   - Minutes On/Off
   - ORtg On/Off (+ differential)
   - DRtg On/Off (+ differential)
   - NetRtg On/Off (+ differential)

   Save as: aces_onoff_2025_manual.csv

2. Run calculate_metrics_from_csv.py:
   - Loads aces_player_box_2025.csv
   - Calculates Four Factors, efficiency metrics
   - Outputs: aces_player_box_metrics_2025.csv

3. Merge everything:
   - Combine box score metrics + on/off splits
   - Create final player profiles dataset

OUTPUT FILES CREATED:
- aces_player_box_2025.csv (filtered raw data)
- aces_top8_rotation_2025.csv (Top 8 list)
- aces_player_summary_2025.csv (all players summary)
- aces_top8_detailed_stats_2025.csv (Top 8 with detailed stats)

TOP 8 PLAYER IDS FOR PBPSTATS:
""")

for _, row in top_8.iterrows():
    print(f"  {row['player_name']}: {row['player_id']}")

print(f"\nCompleted: {pd.Timestamp.now()}")
print("="*70)

LAS VEGAS ACES - TOP 8 ROTATION IDENTIFIER
Input: /content/player_box_2025.parquet
Team: Las Vegas Aces
Season: 2025

[STEP 1] Loading raw player box scores...
✓ Loaded 7,140 rows
  Columns: ['game_id', 'season', 'season_type', 'game_date', 'game_date_time', 'athlete_id', 'athlete_display_name', 'team_id', 'team_name', 'team_location', 'team_short_display_name', 'minutes', 'field_goals_made', 'field_goals_attempted', 'three_point_field_goals_made']...

  Checking team identifiers in data...
  Unique team_ids: [np.int32(3), np.int32(5), np.int32(6), np.int32(8), np.int32(9), np.int32(11), np.int32(14), np.int32(16), np.int32(17), np.int32(18), np.int32(19), np.int32(20), np.int32(129689), np.int32(131246), np.int32(131247)]
  Unique team names: ['Atlanta Dream', 'Chicago Sky', 'Connecticut Sun', 'Dallas Wings', 'Golden State Valkyries', 'Indiana Fever', 'Las Vegas Aces', 'Los Angeles Sparks', 'Minnesota Lynx', 'New York Liberty']...

[STEP 2] Filtering to Las Vegas Aces...
  → By team_i

/tmp/ipython-input-2396038563.py:178: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_lva['minutes_decimal'] = pd.to_numeric(df_lva[minutes_col], errors='coerce').fillna(0)
